lib/src/object_store.rs

```diff
@@ -60,17 +91,94 @@ impl ObjectStores {
         obj_store_key: ObjectStoreKey,
         obj_key: ObjectKey,
         obj: Vec<u8>,
-    ) -> Result<(), ObjectStoreError> {
+        mode: KvInsertMode,
+        generation: Option<u32>,
+        metadata: Option<Vec<u8>>,
+        ttl: Option<std::time::Duration>,
+    ) -> Result<(), KvStoreError> {
+        // manages ttl
+        let existing = self.lookup(obj_store_key.clone(), obj_key.clone());
+
+        if let Some(g) = generation {
+            if let Ok(val) = &existing {
+                if val.generation != g {
+                    return Err(KvStoreError::PreconditionFailed);
+                }
+            }
+        }
+
+        let out_obj = match mode {
+            KvInsertMode::Overwrite => obj,
+            KvInsertMode::Add => {
+                if existing.is_ok() {
+                    // key exists, add fails
+                    return Err(KvStoreError::PreconditionFailed);
+                }
+                obj
+            }
+            KvInsertMode::Append => {
+                let mut out_obj;
+                match existing {
+                    Err(KvStoreError::NotFound) => {
+                        out_obj = obj;
+                    }
+                    Err(_) => return Err(KvStoreError::InternalError),
+                    Ok(v) => {
+                        out_obj = v.body;
+                        out_obj.append(&mut obj.clone());
+                    }
+                }
+                out_obj
+            }
+            KvInsertMode::Prepend => {
+                let mut out_obj;
+                match existing {
+                    Err(KvStoreError::NotFound) => {
+                        out_obj = obj;
+                    }
+                    Err(_) => return Err(KvStoreError::InternalError),
+                    Ok(mut v) => {
+                        out_obj = obj;
+                        out_obj.append(&mut v.body);
+                    }
+                }
+                out_obj
+            }
+        };
+
+        let exp = ttl.map(|t| SystemTime::now() + t);
+
+        let mut obj_val = ObjectValue {
+            body: out_obj,
+            metadata: vec![],
+            metadata_len: 0,
+            generation: SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_nanos() as u32,
+            expiration: exp,
+        };
+
+        // magic number hack to ensure a case for integration tests
+        if obj_val.generation == 1337 {
+            obj_val.generation = 1338;
+        }
```

Comment:
Is this needed for Viceroy? I see it's in the mock app used in the integration tests, but I don't see anything in this PR that needs this.

Reply:
You know what, let me see if I can use this in a unit test, just to validate the failure mode. If not, I'll remove it.

But yeah, otherwise this solely serves the test fixture in the other codebase.

Reply:
new tests do in fact use this special generation for testing the failure case